h/streamer/filter.py

```diff
@@ -113,21 +118,33 @@ def clause_filter(annotation):
         return clause_filter
 
 
-def uni_fold(text):
-    """
-    Return a case-folded and Unicode-normalized copy of ``text``.
+class NormalizedAnnotation:
+    """A normalized and caching version of an annotation.
 
-    This is used to ensure matching of filters against annotations ignores
-    differences in case or different ways of representing the same characters.
-    """
-    # Convert bytes to text
-    if isinstance(text, bytes):
-        text = str(text, "utf-8")
+    This version proxies, normalises and caches the fields used in the filter
+    algorithm to speed up processing. The result can be quite dramatic over
+    a large number of clients.
 
-    # Do not touch other types
-    if not isinstance(text, str):
-        return text
+    The intended use is to wrap an annotation before passing to a filter
+    `match()` method.
+    """
 
-    text = text.lower()
-    text = unicodedata.normalize("NFKD", text)
-    return "".join([c for c in text if not unicodedata.combining(c)])
+    def __init__(self, annotation):
+        self._annotation = annotation
+
+    @property
+    @lru_cache(1)
+    def target_uri(self):
+        return normalize_uri(self._annotation.target_uri)
+
+    @property
+    @lru_cache(1)
+    def references(self):
+        # Ids require no normalisation, so just pass through
+        return self._annotation.references
+
+    @property
+    @lru_cache(1)
+    def id(self):
+        # Ids require no normalisation, so just pass through
```

Comment:
Same question as above, `@lru_cache` is probably not going to be any faster than just executing the method every time.